### PR TITLE
Expose channel count for FixedChannelPool

### PR DESCRIPTION
**Pull request by rockwotj**
Motivation:

We use FixedChannelPool in production, and we believe we have a leak that doesn't return sockets to the pool (but they should be closed), thus blocking us from creating new connections when we need them. I haven't confirmed this yet, but right now I have to resort to reflection to access this field which makes me sad.

Modification:

Expose the acquiredChannelCount field through a getter method.

Result:

Allows introspection of the pool size in FixedChannelPool.

### CODE REVIEW

File: transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java [hunk 1]
```diff
@@ -228,6 +228,11 @@ public void onTimeout(AcquireTask task) {
         this.maxConnections = maxConnections;
         this.maxPendingAcquires = maxPendingAcquires;
     }
+    
+    /** Returns the number of acquired channels that this pool thinks it has. */
+    public int getAcquiredChannelCount() {
+        return acquiredChannelCount;
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: This is not safe as acquiredChannelCount is not atomic.

File: transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java [hunk 2]
```diff
@@ -228,6 +228,11 @@ public void onTimeout(AcquireTask task) {
         this.maxConnections = maxConnections;
         this.maxPendingAcquires = maxPendingAcquires;
     }
+    
+    /** Returns the number of acquired channels that this pool thinks it has. */
+    public int getAcquiredChannelCount() {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: Remove the `get` prefix as we usually not use this prefixes in Netty.

File: transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java (hunk 1 above)
Comment (rockwotj) [reply transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: I see three options, do you have a preference?

1) Make `acquiredChannelCount` atomic
2) Enforce the caller is calling this in the event loop
3) Return a future and call it in the event loop

Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: I guess making acquiredChannelCount atomic is good enough

Comment (rockwotj) [reply transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: Made it atomic.

File: transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java [hunk 3]
```diff
@@ -70,7 +71,7 @@
     private final Queue<AcquireTask> pendingAcquireQueue = new ArrayDeque<AcquireTask>();
     private final int maxConnections;
     private final int maxPendingAcquires;
-    private int acquiredChannelCount;
+    private AtomicInteger acquiredChannelCount = new AtomicInteger();
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: final

Comment (rockwotj) [reply transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: good call, done.

**Comment by normanmaurer** [comment]
@rockwotj can you please squash your commits and sign our ICLA:

https://netty.io/s/icla

Please let me know once done and I will merge this one. Thanks!

**Comment by wsrspirit** [comment]
I think make acquiredChannelCount  atomic can not completely fix this bug,
even though AtomicInteger CAS method is atomic, but the following lines of code are not atomic,
so it can not filly solve this problem 

if (acquiredChannelCount.get() < maxConnections) {
...
acquiredChannelCount.incrementAndGet();
...
}

On the oter side, I hope FixedChannelPool not useing deque(in SimpleChannelPool), which means I can figure out sockets package based on seq num, how can I do?

**Comment by rockwotj** [comment]
This code runs in an single event loop so it doesn't need to be atomic in that situation (note it being a normal `int` before.

**Comment by rockwotj** [comment]
@normanmaurer done.

**Comment by normanmaurer** [comment]
@rockwotj thanks!

@wsrspirit I dont understand your question. If you think there is any issue please open a new issue as this is not related to this change.

**Pull request merged**